storage/stats.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Spencer Kimball (spencer.kimball@gmail.com)
+
+package storage
+
+import (
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/storage/engine"
+)
+
+// A rangeStats encapsulates access to a range's stats. Range
+// statistics are maintained on every range operation using
+// stat increments accumulated via MVCCStats structs. Stats are
+// efficiently aggregated using the engine.Merge operator.
+type rangeStats struct {
+	raftID int64
+	// The following values are cached from the underlying stats.
+	elapsedNanos int64 // nanos since the epoch
+	intentCount  int64
+}
+
+// newRangeStats creates a new instance of rangeStats using the
+// provided engine and range. In particular, the values of elapsed
+// nanos and intent count are pulled from the engine and cached in
+// the struct for efficient processing (i.e. each new merge does
+// not require the values to be read from the engine).
+func newRangeStats(raftID int64, e engine.Engine) (*rangeStats, error) {
```

Comment:
Unfortunately, sometimes you want to read or write directly to the underlying engine, but most times you want to write to the current batch engine, which is created anew for each separate range command execution.